FV3/atmos_cubed_sphere/tools/online_coarse_graining.F90

```diff
@@ -0,0 +1,267 @@
+module online_coarse_graining_mod
+
+  use fv_arrays_mod, only: fv_atmos_type
+  use fms_mod, only: check_nml_error, close_file, open_namelist_file
+  use mpp_domains_mod, only: domain2d, mpp_define_io_domain, mpp_define_mosaic, mpp_get_compute_domain
+  use mpp_mod, only: FATAL, input_nml_file, mpp_error, mpp_npes
+  
+  implicit none
+  private
+
+  public :: block_sum, get_fine_array_bounds, get_coarse_array_bounds, online_coarse_graining_init, weighted_block_average, MODEL_LEVEL
+
+  interface block_sum
+     module procedure block_sum_2d
+  end interface block_sum
+  
+  interface weighted_block_average
+     module procedure weighted_block_average_2d
+     module procedure weighted_block_average_3d_field_2d_weights
+  end interface weighted_block_average
+  
+  ! Global variables for the module, initialized in online_coarse_graining_init
+  type(domain2d) :: coarse_domain
+  integer :: is, ie, js, je, npz
+  integer :: is_coarse, ie_coarse, js_coarse, je_coarse
+  integer :: target_resolution
+  character(len=11) :: MODEL_LEVEL = 'model_level'
+  
+  ! Namelist parameters initialized with default values
+  integer :: coarsening_factor = 8
+  integer :: coarse_io_layout(2) = (/1, 1/)
+  character(len=64) :: coarse_graining_strategy = 'model_level'
+  logical :: do_coarse_graining = .false.
+  
+  namelist /online_coarse_graining_nml/ coarsening_factor, coarse_io_layout, &
+       coarse_graining_strategy, do_coarse_graining
+
+contains
+
+  subroutine online_coarse_graining_init(Atm)
```

Comment:
I think needs to take in `nml_filename`, `layout`, `coarse_graining_attrs`, `npz`, and `bd`.